src/main/java/com/google/devtools/build/lib/bazel/bzlmod/GsonTypeAdapterUtil.java

```diff
@@ -82,47 +82,97 @@ public void write(JsonWriter jsonWriter, ModuleKey moduleKey) throws IOException
         @Override
         public ModuleKey read(JsonReader jsonReader) throws IOException {
           String jsonString = jsonReader.nextString();
-          if (jsonString.equals("<root>")) {
-            return ModuleKey.ROOT;
-          }
-          List<String> parts = Splitter.on('@').splitToList(jsonString);
-          if (parts.get(1).equals("_")) {
-            return ModuleKey.create(parts.get(0), Version.EMPTY);
-          }
-
-          Version version;
           try {
-            version = Version.parse(parts.get(1));
+            return ModuleKey.fromString(jsonString);
           } catch (ParseException e) {
             throw new JsonParseException(
                 String.format("Unable to parse ModuleKey %s version from the lockfile", jsonString),
                 e);
           }
-          return ModuleKey.create(parts.get(0), version);
         }
       };
 
-  // TODO(salmasamy) need to handle "isolated" in module extensions when it is stable
   public static final TypeAdapter<ModuleExtensionId> MODULE_EXTENSION_ID_TYPE_ADAPTER =
       new TypeAdapter<>() {
         @Override
         public void write(JsonWriter jsonWriter, ModuleExtensionId moduleExtId) throws IOException {
-          jsonWriter.value(moduleExtId.getBzlFileLabel() + "%" + moduleExtId.getExtensionName());
+          String extension = moduleExtId.getBzlFileLabel() + "%" + moduleExtId.getExtensionName();
+          if (moduleExtId.getIsolationKey().isEmpty()) {
+            jsonWriter.value(extension);
+          } else {
+            jsonWriter
+                .beginObject()
+                .name("extension")
+                .value(extension)
+                .name("isolation_key")
+                .value(moduleExtId.getIsolationKey().get().toString())
+                .endObject();
+          }
         }
 
         @Override
         public ModuleExtensionId read(JsonReader jsonReader) throws IOException {
+          if (jsonReader.peek() == JsonToken.BEGIN_OBJECT) {
```

Comment:
this is quite a lot of code. could we simplify this by folding everything into a string, something like `//abc.bzl%ext_name[%isolation_key]`?

Reply:
I was thinking about this as well, but then found this to be confusing in its own way as we would essentially concatenate a variable number of up to five distinct pieces of information into a single string. If you prefer it that way I can certainly implement it, I just slightly leaned towards the current solution.

Reply:
actually, we may *have* to concatenate everything here, since a ModuleExtensionId is used as a JSON object key for the extension map. I just remembered that this was why we had a concatenated string for ModuleExtensionId (instead of the normal JSON object representation) in the first place. But then again, your test passed... so I'm not quite sure what's happening here. I wonder what the lockfile actually looks like in that test case.

Reply:
Instead of a map, it has a list of pairs.

Reply:
Should we explore going back to the regular object representation for everything?

Reply:
Are you sure? With a simple local test case, this is what I see in the lockfile with Bazel HEAD:

```json
"moduleExtensions": {
    "//:ext.bzl%ext": {
      "bzlTransitiveDigest": "X8hz9J00k/ncSFM8d3lFoFATXO4NhBFr5xuqfhWiRs0=",
      "envVariables": {},
      "generatedRepoSpecs": {
        "kek": {
          "bzlFile": "@@//:ext.bzl",
          "ruleClassName": "rr",
          "attributes": {"name":"--_main~ext~kek"}
        }
      }
    }
  }
```

Reply:
You're right. What's your preference? Strings, list of pairs or the current solution that's a max of both? Probably not the latter, I suppose.

Reply:
In fact, I wanted to go a bit further and squash things like Starlark's Location into a string (so `"abc.bzl:5:6"` instead of `{"file":"abc.bzl","line":5,"column":6}`. But that would require bumping the lockfile version so I thought to do it later.

Reply:
Makes sense, I can change this tomorrow.

Regarding compressing location: Wouldn't now be a good time given that lockfiles haven't been enabled by default in a release yet?